libs/common/src/main/java/org/opensearch/common/hash/T1ha.java

```diff
@@ -0,0 +1,226 @@
+     * @param length the length of the range
+     * @param seed customized seed
+     * @return hash code
+     */
+    public static long hash(byte[] input, int offset, int length, long seed) {
+        long a = seed;
+        long b = length;
+
+        if (length > 32) {
+            long c = rotateRight(length, s1) + seed;
+            long d = length ^ rotateRight(seed, s1);
+
+            do {
+                long w0 = fetch64(input, offset);
+                long w1 = fetch64(input, offset + 8);
+                long w2 = fetch64(input, offset + 16);
+                long w3 = fetch64(input, offset + 24);
+
+                long d02 = w0 ^ rotateRight(w2 + d, s1);
+                long c13 = w1 ^ rotateRight(w3 + c, s1);
+                c += a ^ rotateRight(w0, s0);
+                d -= b ^ rotateRight(w1, s2);
+                a ^= p1 * (d02 + w3);
+                b ^= p0 * (c13 + w2);
+
+                offset += 32;
+                length -= 32;
+            } while (length >= 32);
+
+            a ^= p6 * (rotateRight(c, s1) + d);
+            b ^= p5 * (rotateRight(d, s1) + c);
+        }
+
+        return h32(input, offset, length, a, b);
+    }
+
+    /**
+     * Computes the hash of up to 32 bytes.
+     * Constants in the switch expression are dense; JVM will use them as indices into a table of
+     * instruction pointers (tableswitch instruction), making lookups really fast.
+     */
+    @SuppressWarnings("fallthrough")
+    private static long h32(byte[] input, int offset, int length, long a, long b) {
+        switch (length) {
+            default:
+                b += mux64(fetch64(input, offset), p4);
+                offset += 8;
+                length -= 8;
+            case 24:
+            case 23:
+            case 22:
+            case 21:
+            case 20:
+            case 19:
+            case 18:
+            case 17:
+                a += mux64(fetch64(input, offset), p3);
+                offset += 8;
+                length -= 8;
+            case 16:
+            case 15:
+            case 14:
+            case 13:
+            case 12:
+            case 11:
+            case 10:
+            case 9:
+                b += mux64(fetch64(input, offset), p2);
+                offset += 8;
+                length -= 8;
+            case 8:
+            case 7:
+            case 6:
+            case 5:
+            case 4:
+            case 3:
+            case 2:
+            case 1:
+                a += mux64(tail64(input, offset, length), p1);
+            case 0:
+                // Final weak avalanche
+                return mux64(rotateRight(a + b, s1), p4) + mix64(a ^ b, p0);
+        }
+    }
+
+    /**
+     * XOR the high and low parts of the full 128-bit product.
+     */
+    private static long mux64(long a, long b) {
+        // Ideally, the following should be used to match the reference implementation:
+        // return Math.unsignedMultiplyHigh(a, b) ^ (a * b);
```

Comment:
I thought about it too but decided against having a bimodal behavior unless there's real benefit for our use-case. We can always converge towards the reference implementation later on as we make no stability guarantees on the returned values (I'll make this more explicit in the Javadocs). I'm still open to this approach though.

On a side note, once we jump the ship to JDK 20 and the SIMD-friendly `t1ha3_atonce` is released, it would be good to port it over with the Vector API. :)

Reply:
> On a side note, once we jump the ship to JDK 20 and the SIMD-friendly t1ha3_atonce is released, it would be good to port it over with the Vector API. :)

We do ship 3.0.0 snapshot builds with JDK-20, and I suspect both `2.x` and `main` would be using JDK-21 in ~September, this is why I think the usage of the new API would become available very soon (also, Apache Lucene uses Project Panama already).

Reply:
> Out of curiosity, I haven't used MethodHandles for switching between implementation. Did you mean the Reflection API?

In some sense, yes, the MethodHandles are under the hood of Reflection APIs in latest JDKs

Reply:
Useful insights! In that case, it should be okay to target unsignedMultiplyHigh for latest JDKs and fallback to multiplyHigh for older ones, like you said. :)

Reply:
Done! :)